%\VignetteIndexEntry{Creating IGV HTML reports with tracktables}
%\VignettePackage{tracktables}
%\VignetteEngine{knitr::knitr}

% To compile this document
% library('knitr'); rm(list=ls()); knit('tracktables.Rnw')

\documentclass[12pt]{article}


<<knitr, echo=FALSE, results="hide">>=
library("knitr")
opts_chunk$set(tidy=FALSE,dev="png",fig.show="hide",
               fig.width=4,fig.height=4.5,
               message=FALSE)
@ 

<<style, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@

<<loadDESeq2, echo=FALSE>>=
library("tracktables")
@


\author{Thomas Carroll$^{1*}$\\[1em] \small{$^{1}$ Bioinformatics Core, MRC Clinical Sciences Centre;} \\ \small{\texttt{$^*$thomas.carroll (at)imperial.ac.uk}}}

\title{Creating IGV HTML reports with tracktables}

\begin{document}

<<include=FALSE>>=
library(knitr)
opts_chunk$set(
concordance=TRUE
)
@


\maketitle

\begin{abstract}
 
 
 Visualising genomics data in genome browsers is a key step in both quality control and the initial interrogation of the hypothesis under investigation. 

 The organisation of large collections of genomics data (such as from large scale high-thoughput sequencing experiments) alongside associated sample or experimental metadata allows for the rapid evaluation of patterns across experimental groups. 

  Broad's Intergrative Genome Viewer (IGV) provides a set of methods to make use of sample metadata when visualising genomics data. As well as assisting in the identifaction of sample types within the genome browser, this sample metadata can be further used in IGV to group, sort and filter samples.
  
  This use of sample metadata, alongside the ability to control IGV through ports, provides the desired framework to rapidly interrogate large cohorts of genomics data once the appropriate file types are built.
  
 The Tracktables package provides a set of tools to build IGV session files from data-frames of sample files and their associated metadata as well as produce IGV linked HTML reports for high-throughput visualisation of sample data in IGV.
 
  \vspace{1em}
  
  \end{abstract}

<<options, results="hide", echo=FALSE>>=
options(digits=3, width=80, prompt=" ", continue=" ")
@

\tableofcontents

\section{Creating IGV sessions and HTML reports using tracktables}

The two main functions within the \texttt{tracktables} package are the \Rfunction{MakeIGVSession()} function for  creating IGV session XMLs and any associated sample metadata files and the \Rfunction{maketracktable()} function to create HTML pages containing sample metadata and interval tables which can be used to control IGV.

\subsection{Creating input files for tracktables}

\texttt{tracktables} functions require the user to provide both a data-frame of metadata information and one of sample file locations to be visualised in IGV. 

These data-frames must both have one column named "SampleName" which contains unique sample IDs. This column will be used to match samples and only samples within both files will be included in the IGV session.

The remaining metadata SampleSheet columns may be user-defined but must all contain column titles. (See example below)

The FileSheet (with file locations) must contain the columns "SampleName", "bam","bigwig" and "interval". These columns may contain NA values when no relavant file is associated to a sample.

Here we create a small example SampleSheet (containing metadata) and FileSheet (containing file locations) from some Histone, RNA polymerase 2 and Ebf ChIP-seq.


<<inputdata, eval=TRUE>>=
library(tracktables)

fileLocations <- system.file("extdata",package="tracktables")

bigwigs <- dir(fileLocations,pattern="*.bw",full.names=TRUE)
intervals <- dir(fileLocations,pattern="*.bed",full.names=TRUE)
bigWigMat <- cbind(gsub("_Example.bw","",basename(bigwigs)),
                   bigwigs)
intervalsMat <- cbind(gsub("_Peaks.bed","",basename(intervals)),
                      intervals)

FileSheet <- merge(bigWigMat,intervalsMat,all=TRUE)
FileSheet <- as.matrix(cbind(FileSheet,NA))
colnames(FileSheet) <- c("SampleName","bigwig","interval","bam")

SampleSheet <- cbind(as.vector(FileSheet[,"SampleName"]),
                     c("EBF","H3K4me3","H3K9ac","RNAPol2"),
                     c("ProB","ProB","ProB","ProB"))
colnames(SampleSheet) <- c("SampleName","Antibody","Species")

@

The SampleSheet can be seen to contain a very small section of metadata for EBF, RNApol2, H3K4me3 and H3K9ac. "SampleName" column contains the unique IDs.

<<quick7, eval=TRUE>>=
head(SampleSheet)
@

The FileSheet contains the "SampleName" column with unique ID which match with those found in the SampleSheet. The remaining columns of "bam","bigwig" and "interval" are also all present.


<<quick99, eval=TRUE>>=
head(FileSheet)
@

Note that not all samples have intervals associated to them and ,here, none of these samples have BAM files associated to them. NA values within the FileSheet will be ignored by \texttt{tracktables} functions.

\subsection{Creating an IGV session XML file}

\texttt{tracktables} can create an IGV session XML and associated sample information file from this SampleSheet and FileSheet.

In addition to the FileSheet and SampleSheet, the \Rfunction{MakeIGVSession()} function requires the location to write to, the filename for the session XML and the genome to be used in IGV (see IGV for details on supported genomes).

<<IGVsessionAndSample, eval=FALSE>>=

MakeIGVSession(FileSheet,igvdirectory=getwd(),"Example","mm9")
@

This creates two files in the current working directory containing the sample information file for IGV ("SampleMetadata.txt") and the session XML itself to be loaded into IGV ("Example.xml").


\subsection{Creating a Tracktable report HTML pages}

As well as producing standalone session XMLs and sample information files, the \texttt{tracktables} package can produce HTML reports which contain metadata and methods to control IGV.

The report structure is made of a main "Tracktables Experiment Report" which houses the metadata from "SampleSheet" and links to open a sample's files in IGV (the sample's bigwig, bam or interval files). All sample files are associated with their relevant sample metadata and grouped together by their unique "SampleName". 

When a sample has an interval file associated to it, the "Tracktables Experiment Report" also contains a link to a further sample specific "Tracktables Interval Report". This interval report contains a table of interval locations, any metadata associated with intervals and further links to focus IGV on an interval's genomic location. 

<<quick5, eval=FALSE>>=
  HTMLreport <- maketracktable(FileSheet,SampleSheet,
                               "IGVExample.html",
                               basedirectory=getwd(),
                               "mm9")
@

The maketracktables() function, here, creates an HTML report "IGVExample.html" (the Tracktables Experiment Report) in the current working directory alongside the relevant sample IGV session XMLs, Tracktables Interval Reports and sample information files. The function also returns the XML doc to allow the user to add further customisation to the main report.

Since \texttt{tracktables} uses relative paths to communicate with IGV, in practice the creation of Tracktable's reports in a new directory, alongside any files to display, is advised. This allows for the report to be high portable and so delivered to the user as a functional unit to use with IGV.

\end{document}
